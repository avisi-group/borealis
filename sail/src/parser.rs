//! Sail source file parsing

use {
    crate::{error::Error, RT},
    ocaml::{
        interop::{BoxRoot, ToOCaml},
        List, Value,
    },
    once_cell::sync::Lazy,
};

ocaml::import! {
    fn internal_type_check_initial_env() -> Value;

    // val load_files : ?check:bool -> (Arg.key * Arg.spec * Arg.doc) list -> Type_check.Env.t -> string list -> (string * Type_check.tannot ast * Type_check.Env.t)
    fn internal_process_file_load_files(check: bool, options: List<Value>, env: Value, files: List<BoxRoot<String>>) -> (String, Value, Value);
}

/// Parses supplied Sail files and returns the AST
///
/// From Sail internal docs:
///
/// This function parses all the files passed to Sail, and then concatenates
/// their ASTs. The pre-processor is then run, which evaluates `$directive`
/// statements in Sail, such as
///
/// ```sail
/// $include <prelude.sail>
/// ```
///
/// Unlike the C pre-processor the Sail pre-processor operates over actual
/// Sail ASTs rather than strings. This can recursively include other
/// files into the AST, as well as add/remove parts of the AST with
/// `$ifdef` etc. Directives that are used are preserved in the AST, so
/// they also function as a useful way to pass auxiliary information to
/// the various Sail backends.
///
/// The initial check mentioned above is then run to desugar the AST, and
/// then the type-checker is run which produces a fully type-checked
/// AST. Type annotations are attached to every node (for which an
/// annotation makes sense) using the aux constructors.
///
/// After type-checking the Sail scattered definitions are de-scattered
/// into single functions.
pub fn load_files(files: Vec<String>) -> Result<(), Error> {
    Lazy::force(&RT);

    let env = unsafe { internal_type_check_initial_env(&*RT.lock())? };

    let mut file_list = List::empty();

    for file in files {
        let file_rooted: BoxRoot<String> = file.to_boxroot(&mut *RT.lock());
        file_list = unsafe { file_list.add(&*RT.lock(), &file_rooted) };
    }

    let (output_name, ast, type_envs) = unsafe {
        internal_process_file_load_files(&*RT.lock(), false, List::empty(), env, file_list)?
    };

    dbg!((output_name, ast, type_envs));

    Ok(())
}

#[cfg(test)]
mod tests {
    use {
        crate::parser::load_files,
        std::{env, path::PathBuf},
    };

    #[test]
    fn load_files_empty() {
        load_files(vec![]).unwrap();
    }

    #[test]
    fn load_files_prelude() {
        let path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap())
            .join("examples/prelude.sail")
            .to_string_lossy()
            .to_string();

        load_files(vec![path.clone()]).unwrap();
    }
}
