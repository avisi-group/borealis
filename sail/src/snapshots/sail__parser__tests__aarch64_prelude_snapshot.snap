---
source: sail/src/parser.rs
expression: "parser().parse(Stream::from_iter(source.len()..source.len() + 1,\n            tokens.into_iter())).unwrap()"
---
[]
